app/controllers/api/v1/base_controller.rb

```diff
@@ -3,14 +3,104 @@
 module Api
   module V1
     class BaseController < ApiController
-      rescue_from ActionController::ParameterMissing do |e|
+      class TypeError < StandardError; end
+      class IDMismatchError < StandardError; end
+
+      rescue_from StandardError do |e|
         logger.error e.message
-        render jsonapi: {}, status: :bad_request
+        logger.error e.backtrace.join("\n")
+        render_error(I18n.t('api.core.errors.general.title'),
+                     I18n.t('api.core.errors.general.detail'),
+                     :bad_request)
+      end
+
+      rescue_from TypeError do
+        render_error(I18n.t('api.core.errors.type.title'),
+                     I18n.t('api.core.errors.type.detail'),
+                     :bad_request)
+      end
+
+      rescue_from IDMismatchError do
+        render_error(I18n.t('api.core.errors.id_mismatch.title'),
+                     I18n.t('api.core.errors.id_mismatch.detail'),
+                     :bad_request)
+      end
+
+      rescue_from ActionController::ParameterMissing do |e|
+        render_error(
+          I18n.t('api.core.errors.parameter.title'), e.message, :bad_request
+        )
       end
 
       rescue_from ActiveRecord::RecordNotFound do |e|
-        logger.error e.message
-        render jsonapi: {}, status: :not_found
+        render_error(
+          I18n.t('api.core.errors.record_not_found.title'),
+          I18n.t('api.core.errors.record_not_found.detail',
+                 model: e.model,
+                 id: e.id),
+          :not_found
+        )
+      end
+
+      rescue_from ActiveRecord::RecordInvalid do |e|
+        render_error(
+          I18n.t('api.core.errors.validation.title'), e.message, :bad_request
+        )
+      end
+
+      private
+
+      def render_error(title, message, status)
+        logger.error message
+        render json: {
+          errors: [
+            {
+              id: request.uuid,
+              status: Rack::Utils.status_code(status),
+              title: title,
+              detail: message
+            }
+          ]
+        }, status: status
+      end
+
+      def permission_error(klass, mode)
```

Comment:
Wouldn't it be more streamlined if permission error is another sub-class of `StandardError` (with the 2 parameters, `klass` and `mode`). So whenever you want to render error from within descendants of `BaseController`, you always just use `raise`, instead of 2 ways (`raise` for majority of errors, and `permission_error` for permission errors). What do you think?

Reply:
Done